### ISSUE TITLE
`mut-mut` may remove required lifetime

### ISSUE THREAD

**Issue by matthiaskrgr** [rust-lang/rust-clippy#15838]
Using the following flags

`--force-warn clippy::mut-mut`

this code:

```rust
//@ edition:2021
//@ check-pass

#![feature(type_alias_impl_trait)]
#![feature(closure_lifetime_binder)]

use std::future::Future;

trait AsyncFn<I, R>: FnMut(I) -> Self::Fut {
    type Fut: Future<Output = R>;
}

impl<F, I, R, Fut> AsyncFn<I, R> for F
where
    Fut: Future<Output = R>,
    F: FnMut(I) -> Fut,
{
    type Fut = Fut;
}

async fn call<C, R, F>(mut ctx: C, mut f: F) -> Result<R, ()>
where
    F: for<'a> AsyncFn<&'a mut C, Result<R, ()>>,
{
    loop {
        match f(&mut ctx).await {
            Ok(val) => return Ok(val),
            Err(_) => continue,
        }
    }
}

trait Cap<'a> {}
impl<T> Cap<'_> for T {}

fn works(ctx: &mut usize) {
    let mut inner = 0;

    type Ret<'a, 'b: 'a> = impl Future<Output = Result<usize, ()>> + 'a + Cap<'b>;

    let callback = for<'a, 'b> |c: &'a mut &'b mut usize| -> Ret<'a, 'b> {
        inner += 1;
        async move {
            let _c = c;
            Ok(1usize)
        }
    };
    call(ctx, callback);
}

fn doesnt_work_but_should(ctx: &mut usize) {
    let mut inner = 0;

    type Ret<'a, 'b: 'a> = impl Future<Output = Result<usize, ()>> + 'a + Cap<'b>;

    call(ctx, for<'a, 'b> |c: &'a mut &'b mut usize| -> Ret<'a, 'b> {
        inner += 1;
        async move {
            let _c = c;
            Ok(1usize)
        }
    });
}

fn main() {}

```

caused the following diagnostics:

```
    Checking _supertrait-hint-cycle v0.1.0 (/tmp/icemaker_global_tempdir.paNtiJF6jVLU/icemaker_clippyfix_tempdir.4K5QhGvIQOZc/_supertrait-hint-cycle)
warning: a type of form `&mut &mut _`
  --> src/main.rs:41:36
   |
41 |     let callback = for<'a, 'b> |c: &'a mut &'b mut usize| -> Ret<'a, 'b> {
   |                                    ^^^^^^^^^^^^^^^^^^^^^ help: remove the extra `&mut`: `&'b mut usize`
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#mut_mut
   = note: requested on the command line with `--force-warn clippy::mut-mut`

warning: a type of form `&mut &mut _`
  --> src/main.rs:56:31
   |
56 |     call(ctx, for<'a, 'b> |c: &'a mut &'b mut usize| -> Ret<'a, 'b> {
   |                               ^^^^^^^^^^^^^^^^^^^^^ help: remove the extra `&mut`: `&'b mut usize`
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#mut_mut

warning: `_supertrait-hint-cycle` (bin "_supertrait-hint-cycle") generated 2 warnings
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.31s

```

However after applying these diagnostics, the resulting code:

```rust
//@ edition:2021
//@ check-pass

#![feature(type_alias_impl_trait)]
#![feature(closure_lifetime_binder)]

use std::future::Future;

trait AsyncFn<I, R>: FnMut(I) -> Self::Fut {
    type Fut: Future<Output = R>;
}

impl<F, I, R, Fut> AsyncFn<I, R> for F
where
    Fut: Future<Output = R>,
    F: FnMut(I) -> Fut,
{
    type Fut = Fut;
}

async fn call<C, R, F>(mut ctx: C, mut f: F) -> Result<R, ()>
where
    F: for<'a> AsyncFn<&'a mut C, Result<R, ()>>,
{
    loop {
        match f(&mut ctx).await {
            Ok(val) => return Ok(val),
            Err(_) => continue,
        }
    }
}

trait Cap<'a> {}
impl<T> Cap<'_> for T {}

fn works(ctx: &mut usize) {
    let mut inner = 0;

    type Ret<'a, 'b: 'a> = impl Future<Output = Result<usize, ()>> + 'a + Cap<'b>;

    let callback = for<'a, 'b> |c: &mut &'b mut usize| -> Ret<'a, 'b> {
        inner += 1;
        async move {
            let _c = c;
            Ok(1usize)
        }
    };
    call(ctx, callback);
}

fn doesnt_work_but_should(ctx: &mut usize) {
    let mut inner = 0;

    type Ret<'a, 'b: 'a> = impl Future<Output = Result<usize, ()>> + 'a + Cap<'b>;

    call(ctx, for<'a, 'b> |c: &mut &'b mut usize| -> Ret<'a, 'b> {
        inner += 1;
        async move {
            let _c = c;
            Ok(1usize)
        }
    });
}

fn main() {}

```
no longer compiled:

```
    Checking _supertrait-hint-cycle v0.1.0 (/tmp/icemaker_global_tempdir.paNtiJF6jVLU/icemaker_clippyfix_tempdir.4K5QhGvIQOZc/_supertrait-hint-cycle)
error[E0637]: `&` without an explicit lifetime name cannot be used here
  --> src/main.rs:41:36
   |
41 |     let callback = for<'a, 'b> |c: &mut &'b mut usize| -> Ret<'a, 'b> {
   |                                    ^ explicit lifetime name needed here

error[E0637]: `&` without an explicit lifetime name cannot be used here
  --> src/main.rs:56:31
   |
56 |     call(ctx, for<'a, 'b> |c: &mut &'b mut usize| -> Ret<'a, 'b> {
   |                               ^ explicit lifetime name needed here

error: implementation of `std::ops::FnOnce` is not general enough
  --> src/main.rs:43:9
   |
43 | /         async move {
44 | |             let _c = c;
45 | |             Ok(1usize)
46 | |         }
   | |_________^ implementation of `std::ops::FnOnce` is not general enough
   |
   = note: closure with signature `for<'b, 'a> fn(&'2 mut &'b mut usize) -> works::Ret<'a, 'b>` must implement `std::ops::FnOnce<(&'1 mut &mut usize,)>`, for any lifetime `'1`...
   = note: ...but it actually implements `std::ops::FnOnce<(&'2 mut &mut usize,)>`, for some specific lifetime `'2`

error: implementation of `std::ops::FnOnce` is not general enough
  --> src/main.rs:58:9
   |
58 | /         async move {
59 | |             let _c = c;
60 | |             Ok(1usize)
61 | |         }
   | |_________^ implementation of `std::ops::FnOnce` is not general enough
   |
   = note: closure with signature `for<'b, 'a> fn(&'2 mut &'b mut usize) -> doesnt_work_but_should::Ret<'a, 'b>` must implement `std::ops::FnOnce<(&'1 mut &mut usize,)>`, for any lifetime `'1`...
   = note: ...but it actually implements `std::ops::FnOnce<(&'2 mut &mut usize,)>`, for some specific lifetime `'2`

For more information about this error, try `rustc --explain E0637`.
error: could not compile `_supertrait-hint-cycle` (bin "_supertrait-hint-cycle") due to 4 previous errors
warning: build failed, waiting for other jobs to finish...
error: could not compile `_supertrait-hint-cycle` (bin "_supertrait-hint-cycle" test) due to 4 previous errors

```

Version:
```
rustc 1.92.0-nightly (4a54b26d3 2025-10-07)
binary: rustc
commit-hash: 4a54b26d30dac43778afb0e503524b763fce0eee
commit-date: 2025-10-07
host: x86_64-unknown-linux-gnu
release: 1.92.0-nightly
LLVM version: 21.1.2

```

**Comment by ada4a** [comment]
But the version "after" contains `&mut &'b mut usize`, not `&'b mut usize` as the lint suggested?

Granted, removing the `&mut` also doesn't compile, but that's due to the type constraints of the surrounding code. Since it seems to be a test for some complicated stuff, it probably makes sense to silence the lint here, but otherwise it'd probably be best to refactor the surroundings so that they don't require a `&mut &mut`
